notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,458 @@
+    "P\\ [\\ \\text{recidivism predicted}\\ |\\ \\text{African-American, no recidivism}\\ ] = P\\ [\\ \\text{recidivism predicted}\\ |\\ \\text{Caucasian, no recidivism}\\ ], \\text{e.g. } 0.15\n",
+    "$$\n",
+    "\n",
+    "but that also means that men (and women) of different subgroup based on training labels don't necessarily have parity:\n",
+    "\n",
+    "$$\n",
+    "P[\\text{recidivism predicted} | \\text{African-American, recidivism}] = 0.95 \\neq 0.15 = P[\\text{recidivism predicted} | \\text{African-American, no ecidivism}]\n",
+    "$$\n",
+    "\n",
+    "Assessing which disparity metric is indeed fair varies by application scenario."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# This wrapper around the unconstrained estimator serves the purpose of mapping the predict\n",
+    "# method to predict_proba so that we can use real values to get more accurate estimates.\n",
+    "class LogisticRegressionAsRegression:\n",
+    "    def __init__(self, logistic_regression_estimator):\n",
+    "        self.logistic_regression_estimator = logistic_regression_estimator\n",
+    "    \n",
+    "    def fit(self, X, y):\n",
+    "        self.logistic_regression_estimator.fit(X, y)\n",
+    "    \n",
+    "    def predict(self, X):\n",
+    "        # use predict_proba to get real values instead of 0/1, select only prob for 1\n",
+    "        scores = self.logistic_regression_estimator.predict_proba(X)[:,1]\n",
+    "        return scores"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fairlearn.postprocessing import ThresholdOptimizer\n",
+    "from copy import deepcopy\n",
+    "\n",
+    "unconstrained_predictor_wrapper = LogisticRegressionAsRegression(unconstrained_predictor)\n",
+    "postprocessed_predictor_EO = ThresholdOptimizer(\n",
+    "    unconstrained_predictor=unconstrained_predictor_wrapper,\n",
+    "    constraints=\"equalized_odds\",\n",
+    "    plot=True)\n",
+    "\n",
+    "postprocessed_predictor_EO.fit(X_train, y_train, sensitive_features=sensitive_features_train)\n",
+    "\n",
+    "fairness_aware_predictions_EO_train = postprocessed_predictor_EO.predict(X_train, sensitive_features=sensitive_features_train)\n",
+    "fairness_aware_predictions_EO_test = postprocessed_predictor_EO.predict(X_test, sensitive_features=sensitive_features_test)\n",
+    "\n",
+    "show_proportions(X_train, sensitive_features_train, fairness_aware_predictions_EO_train, y_train, description=\"equalized odds with postprocessed model on training data:\", plot_row_index=1)\n",
+    "show_proportions(X_test, sensitive_features_test, fairness_aware_predictions_EO_test, y_test, description=\"equalized odds with postprocessed model on test data:\", plot_row_index=2)"
+   ]
```

Comment:
This is another wall of plots which aren't explained at all.

Reply:
While I understand that there are a few plots, I don't agree with "aren't explained at all". It says very clearly above each plot what it is about. If you feel like I should remove the plots related to training data I can understand that, although that's kind of the whole point: you train on training data, so that's where your constraints are enforced. For test data they *may* generalize and hopefully they do, but you can't say for sure. That's why both training and test data plots are relevant.

We can add some more explicit description of the plots if you find that helpful, but it's pretty much just illustrative of the description of the constraints above.

Reply:
I commented out the training plots, so it's fairly obvious now I think